R/functions-MSnExp.R

```diff
@@ -355,6 +355,222 @@ removeReporters_MSnExp <- function(object, reporters = NULL,
     return(object)
 }
 
+#' @title Estimate m/z scattering in consecutive scans
+#'
+#' @description
+#'
+#' Estimate scattering of m/z values (due to technical, instrument specific
+#' noise) for the same ion in consecutive scans of a LCMS experiment.
+#'
+#' @details
+#'
+#' The m/z values of the same ions in consecutive scans (spectra) of a LCMS run
+#' will not be identical. This random noise is expected to be smaller than the
+#' resolution of the MS instrument. The distribution of differences of m/z
+#' values from neighboring spectra is thus expected to be (at least) bi-modal
+#' with the first peak representing the above described random variation and
+#' the second (or largest) peak the m/z resolution. The m/z value of the first
+#' local minimum between these first two peaks in the distribution is returned
+#' as the *m/z scattering*.
+#'
+#' @param x `MSnExp` or `OnDiskMSnExp` object.
+#'
+#' @param halfWindowSize `integer(1)` defining the half window size for the
+#'     moving window to combine consecutive spectra.
+#' 
+#' @author Johannes Rainer
+#'
+#' @md
+#'
+#' @seealso [estimateMzResolution()] for the function to estimate a
+#'     profile-mode spectrum's m/z resolution from it's data.
+#'
+#' @examples
+#' 
+#' library(MSnbase)
+#' library(msdata)
+#' ## Load a profile-mode LC-MS data file
+#' f <- dir(system.file("sciex", package = "msdata"), full.names = TRUE)[1]
+#' im <- readMSData(f, mode = "inMem", msLevel = 1L)
+#'
+#' res <- estimateMzScattering(im)
+#'
+#' ## Plot the distribution of estimated m/z scattering
+#' plot(density(unlist(res)))
+#'
+#' ## Compare the m/z resolution and m/z scattering of the spectrum with the
+#' ## most peaks
+#' idx <- which.max(unlist(spectrapply(im, peaksCount)))
+#'
+#' res[[idx]]
+#' abline(v = res[[idx]], lty = 2)
+#' estimateMzResolution(im[[idx]])
+#' ## As expected, the m/z scattering is much lower than the m/z resolution.
+estimateMzScattering <- function(x, halfWindowSize = 1L) {
+    if (!is(x, "MSnExp"))
+        stop("'x' should be a 'MSnExp' object")
+    res <- lapply(split(spectra(x), fromFile(x)), function(z) {
+        len_z <- length(z)
+        mzs <- vector("list", len_z)
+        for (i in seq_along(z)) {
+            mzs[[i]] <- .estimate_mz_scattering(
+                sort(unlist(lapply(z[max(1, i - halfWindowSize):
```

Comment:
Didn't know `embed` so far. Looks interesting. If it's OK for you I would however like to keep the current code. It has fewer lines and one loop less (`embed` uses also an internal `for` loop). Let me know what you think.

Reply:
I am fine with your current solution. JFR `embed` is vectorised and doesn't use a `for` loop if the input is a simple vector (and not a `matrix`): https://github.com/wch/r-source/blob/af7f52f70101960861e5d995d3a4bec010bc89e6/src/library/stats/R/embed.R#L34-L41

My suggested implementation with `embed` doesn't improve runtime or anything else (instead it would cause a little bit more memory usage because all windows are precalculated instead of just the current one). So please keep your approach.

Reply:
Thanks for the clarification. I would keep then my implementation - I find it easier to read 😉